### PR TITLE
Improved unit tests structure for ICS-721

### PR DESCRIPTION
**Pull request by humanalgorithm**
- Now covering all functions with unit tests, 30+ tests
- Refactor unit tests into ibc_test.rs and contract_test.rs
- Helper functions for send and receive sg721
- Pass and fail test cases for each endpoint

### CODE REVIEW

**Comment by shanev** [comment]
@humanalgorithm You can run the linter locally with `cargo clippy --all-targets -- -D warnings`.

File: contracts/ics721/src/contract_test.rs [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+        let sender_address: Addr = Addr::unchecked("wasm1fucynrfkrt684pm8jrt8la5h2csvs5cnldcgqc");
+        let result = execute_transfer(deps.as_mut(), mock_env(), transfer_msg, sender_address);
+        let expected_result = [
+            Attribute {
+                key: "action".into(),
+                value: "transfer".into(),
+            },
+            Attribute {
+                key: "sender".into(),
+                value: "wasm1fucynrfkrt684pm8jrt8la5h2csvs5cnldcgqc".into(),
```
Comment (yubrew) [inline contracts/ics721/src/contract_test.rs:378]: why not use var reference, `sender_address.clone()`?

File: contracts/ics721/src/contract_test.rs [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+            channel: "channel-3".to_string(),
+            class_id: "abc/123/collection-addr".to_string(),
+            class_uri: Some("abc/456/collection-addr".to_string()),
+            token_ids: vec!["1".to_string(), "2".to_string(), "3".to_string()],
+            token_uris: vec![
+                "https://metadata-url.com/my-metadata1".to_string(),
+                "https://metadata-url.com/my-metadata2".to_string(),
+                "https://metadata-url.com/my-metadata3".to_string(),
+            ],
+            remote_address: "stars1zedxv25ah8fksmg2lzrndrpkvsjqgk4zt5ff7n".to_string(),
```
Comment (yubrew) [inline contracts/ics721/src/contract_test.rs:366]: what if you define remote_address somewhere else so it's abstracted away

**Comment by yubrew** [comment]
Looks like a good set of tests. A few optional comments on style.

File: contracts/ics721/src/contract.rs [hunk 3]
```diff
@@ -149,22 +153,24 @@ fn query_list(deps: Deps) -> StdResult<ListChannelsResponse> {
...
             let (class_id_token_id, _) = r?;
-            // TODO: extract class_id out of Vec<u8>
-            // https://github.com/public-awesome/contracts/issues/60
-            Ok(class_id_token_id)
+            Ok(class_id_token_id.0)
         })
-        // TODO: filter duplicates
         .collect();
 
+    let mut class_ids_resp = _class_ids.unwrap();
```
Comment (shanev) [inline contracts/ics721/src/contract.rs:165]: I would unwrap the response with `?` after `collect()`, and remove this `unwrap()` so the error can be handled.

File: contracts/ics721/src/contract_test.rs [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+                value: "1,2,3".into(),
+            },
+        ];
+        let result_clone = result.unwrap();
+        assert_eq!(result_clone.attributes, expected_result);
+        assert_eq!(result_clone.messages.len(), 1);
+    }
+
+    #[test]
+    fn test_execute_receive_payment_fail() {
```
Comment (shanev) [inline contracts/ics721/src/contract_test.rs:463]: Should this be `receive_payment` or `receive_nft`?

**Review by shanev — CHANGES_REQUESTED** [review]
Amazing job! Just a few comments.

File: contracts/ics721/src/contract_test.rs (hunk 4 above)
Comment (humanalgorithm) [reply contracts/ics721/src/contract_test.rs:463]: Changed the name to `test_execute_receive_nonpayable_fail`

the function its testing is below, note the nonpayable:
```
pub fn execute_receive(
  ...
) -> Result<Response, ContractError> {
    nonpayable(&info)?;
```
Then the error: 
```
pub fn nonpayable(info: &MessageInfo) -> Result<(), PaymentError> {
      ...
        Err(PaymentError::NonPayable {})
}
```

File: contracts/ics721/src/contract.rs (hunk 3 above)
Comment (humanalgorithm) [reply contracts/ics721/src/contract.rs:165]: Now doing a match: 

```
    let class_ids_resp = _class_ids;
    match class_ids_resp {
        Ok(mut class_id_vec) => Ok(ChannelResponse {
            info,
            class_ids: {
                class_id_vec.sort();
                class_id_vec.dedup();
                class_id_vec
            },
        }),
        Err(msg) => Err(StdError::GenericErr { msg: msg.to_string() })
 ```

**Review by yubrew — APPROVED** [review]
solid pr

**Review by shanev — APPROVED** [review]

**Pull request merged**